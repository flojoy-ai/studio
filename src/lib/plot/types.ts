export interface Drawable {
  draw: () => void;
}
