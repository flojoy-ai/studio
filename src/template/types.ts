export interface SomeObj {
    id: string,
    propertyX: string;
    propertyY: string;
}