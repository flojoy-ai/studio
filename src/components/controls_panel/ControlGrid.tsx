import "react-grid-layout/css/styles.css";
import { Responsive, WidthProvider } from "react-grid-layout";
import { useFlowChartState } from "../../hooks/useFlowChartState";
import "./Controls.css";
import "../../App.css";
import ControlComponent from "./controlComponent";
import { useEffect } from "react";

const ResponsiveGridLayout = WidthProvider(Responsive);

// const atomLayouts = atomWithImmer<Layout[] | undefined>(undefined);

export interface ControlProps {
  theme: any;
  isEditMode: any;
  results: any;
  updateCtrlValue: any;
  attachParam2Ctrl: any;
  rmCtrl: any;
  setCurrentInput: any;
  setOPenEditModal: any;
}

export default function ControlGrid({
  controlProps,
}: {
  controlProps: ControlProps;
}) {
  // const [layout, setLayoutI] = useAtom(atomLayouts);
  const { ctrlsManifest, gridLayout, setGridLayout } = useFlowChartState();
  const { isEditMode } = controlProps;

  // const setLayout = useCallback(
  //   (layouts: Layout[]) => {
  //     console.log("setting layouts to:", layouts);
  //     setLayoutI(layouts);
  //   },
  //   [setLayoutI]
  // );

  // if currently the layouts is undefined, place the controls automatically
  // useEffect(() => {
  //   if (!layout) {
  //     let x = 0;
  //     const ctrlLayouts = ctrlsManifest.map((ctrl, i) => {
  //       const w =  2;//ctrl.label === 'output' ? 40 : 20;
  //       const layoutParams: Layout = {
  //         i: ctrl.name + i,
  //         x,
  //         y: 0,
  //         w,
  //         h: 2,
  //         minH:ctrl.minHeight,
  //         minW:ctrl.minWidth,
  //         static: isEditMode,
  //       };
  //       // x += w;
  //       return layoutParams;
  //     });
  //     console.log('ctrlLayouts: ', ctrlLayouts)
  //     setLayout(ctrlLayouts);
  //   }
  // }, [ctrlsManifest, isEditMode, layout, setLayout]);

  // console.log("ctrlsManifest:", ctrlsManifest);
  // console.log("layouts:", gridLayout);
  useEffect(()=>{
    if(isEditMode){
      setGridLayout(prev=> prev.map(layout=> ({...layout, static:false})))
    } else {
      setGridLayout(prev=> prev.map(layout=> ({...layout, static:true})))
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  },[isEditMode])

  return (
    <ResponsiveGridLayout
      className="layout"
      layouts={{ lg: gridLayout, md: gridLayout, sm: gridLayout }}
      //   breakpoints={{ lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }}
      cols={{ lg: 8, md: 8, sm: 6, xs: 4, xxs: 2 }}
      // rowHeight={130}
      onLayoutChange={(currentLayout, allLayout) => {
        setGridLayout(currentLayout);
      }}
    >
      {ctrlsManifest.map((ctrl, i) => {
        if (ctrl.hidden && !isEditMode) {
          return (
            <div
              key={ctrl.id}
              data-grid={{
                ...gridLayout.find((l) => l.i === ctrl.id)
              }}
              style={{
                display: 'none'
              }}
              data-cy="ctrl-grid-item"
            />
          );
        }
        return (
          <div
            key={ctrl.id}
            data-grid={{
              ...gridLayout.find((l) => l.i === ctrl.id),
              static: !isEditMode,
            }}
            style={{
              ...(controlProps.theme === "dark" && {
                backgroundColor: "#191919",
              }),
              borderRadius:'16px'
              // minHeight:ctrl.minHeight, minWidth:ctrl.minWidth
            }}
            data-cy="ctrl-grid-item"
          >
            <Control
              key={ctrl.id}
              controlProps={controlProps}
              ctrl={ctrl}
              ctrlIndex={i}
            />
          </div>
        );
      })}
    </ResponsiveGridLayout>
  );
}

function Control({
  controlProps,
  ctrl,
  ctrlIndex,
}: {
  controlProps: ControlProps;
  ctrl;
  ctrlIndex: number;
}) {
  const {
    isEditMode,
    theme,
    results,
    updateCtrlValue,
    attachParam2Ctrl,
    rmCtrl,
    setCurrentInput,
    setOPenEditModal,
  } = controlProps;

  return (
    <div
      className={isEditMode ? "ctrl-input" : ""}
      style={{
        width: "100%",
        height: "100%",
        display: "flex",
        flexDirection: "column",
        borderRadius:'16px',
       backgroundColor: theme === "dark" ? "#14131361" : '#58454517',
      }}
    >
      {isEditMode ? (
        <ControlComponent
          ctrlObj={ctrl}
          theme={theme}
          results={results}
          updateCtrlValue={updateCtrlValue}
          attachParam2Ctrl={attachParam2Ctrl}
          rmCtrl={rmCtrl}
          setCurrentInput={setCurrentInput}
          setOPenEditModal={setOPenEditModal}
        />
      ) : ctrl.hidden ? null : (
        <ControlComponent
          ctrlObj={ctrl}
          theme={theme}
          results={results}
          updateCtrlValue={updateCtrlValue}
          attachParam2Ctrl={attachParam2Ctrl}
          rmCtrl={rmCtrl}
          setCurrentInput={setCurrentInput}
          setOPenEditModal={setOPenEditModal}
        />
      )}
    </div>
  );
}
