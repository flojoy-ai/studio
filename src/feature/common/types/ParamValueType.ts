export type ParamValueType =
  | "string"
  | "float"
  | "int"
  | "array"
  | "boolean"
  | "select"
  | "node_reference"
  | "unknown";
