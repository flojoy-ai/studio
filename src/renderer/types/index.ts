export type {
  ALL_DC_TYPE,
  DataContainer,
  DataContainerType,
} from "./dataContainer";
export type { CustomNodeProps, BlockData } from "./node";
export type {
  DataContainer2PlotlyProps,
  OverridePlotData,
  PlotProps,
} from "./plotly";
