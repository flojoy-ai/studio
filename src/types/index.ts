export type {
  ALL_DC_TYPE,
  DataContainer,
  DataContainerType,
} from "./dataContainer";
export type { CustomNodeProps, ElementsData } from "./node";
export type {
  DataContainer2PlotlyProps,
  OverridePlotData,
  PlotProps,
} from "./plotly";
