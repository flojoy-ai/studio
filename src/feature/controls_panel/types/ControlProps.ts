import { ResultsType } from "@src/feature/results_panel/types/ResultsType";
import { CtlManifestType } from "@src/hooks/useFlowChartState";
import { Dispatch, SetStateAction } from "react";
import { CtrlOptionValue } from "./ControlOptions";

export interface ControlProps {
  isEditMode: boolean;
  results: ResultsType;
  updateCtrlValue: (val: string, ctrl: CtlManifestType) => void;
  attachParamsToCtrl: (param: CtrlOptionValue, ctrl: CtlManifestType) => void;
  removeCtrl: (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    ctrl: CtlManifestType
  ) => void;
  setCurrentInput: Dispatch<
    SetStateAction<
      | (CtlManifestType & {
          index: number;
        })
      | undefined
    >
  >;
  setOpenEditModal: Dispatch<SetStateAction<boolean>>;
}
