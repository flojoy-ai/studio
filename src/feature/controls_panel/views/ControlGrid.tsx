import { useEffect } from "react";
import { Responsive, WidthProvider } from "react-grid-layout";
import "react-grid-layout/css/styles.css";
import "@src/App.css";
import { useFlowChartState } from "@src/hooks/useFlowChartState";
import "../style/Controls.css";
import { ControlProps } from "../types/ControlProps";
import Control from "./Control";

const ResponsiveGridLayout = WidthProvider(Responsive);

export default function ControlGrid({
  controlProps,
}: {
  controlProps: ControlProps;
}) {
  const { ctrlsManifest, gridLayout, setGridLayout } = useFlowChartState();
  const { isEditMode } = controlProps;

  useEffect(() => {
    if (isEditMode) {
      setGridLayout((prev) =>
        prev.map((layout) => ({ ...layout, static: false }))
      );
    } else {
      setGridLayout((prev) =>
        prev.map((layout) => ({ ...layout, static: true }))
      );
    }
  }, [isEditMode, setGridLayout]);

  return (
    <ResponsiveGridLayout
      className="layout"
      layouts={{ lg: gridLayout, md: gridLayout, sm: gridLayout }}
      cols={{ lg: 8, md: 8, sm: 6, xs: 4, xxs: 2 }}
      onLayoutChange={(currentLayout, allLayout) => {
        setGridLayout(currentLayout);
      }}
    >
      {ctrlsManifest.map((ctrl, i) => {
        if (ctrl.hidden && !isEditMode) {
          return (
            <div
              key={ctrl.id}
              data-grid={{
                ...gridLayout.find((l) => l.i === ctrl.id),
              }}
              style={{
                display: "none",
              }}
            />
          );
        }

        return (
          <div
            key={ctrl.id}
            data-grid={{
              ...gridLayout.find((l) => l.i === ctrl.id),
              static: !isEditMode,
            }}
            style={{
              ...(controlProps.theme === "dark" && {
                backgroundColor: "#191919",
              }),
              borderRadius: "16px",
            }}
          >
            <Control key={ctrl.id} controlProps={controlProps} ctrl={ctrl} />
          </div>
        );
      })}
    </ResponsiveGridLayout>
  );
}
