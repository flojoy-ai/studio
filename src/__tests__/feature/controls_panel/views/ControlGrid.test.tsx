import { render } from "@testing-library/react";
import ControlGrid from "@src/feature/controls_panel/views/ControlGrid";
import { ControlProps } from "@src/feature/controls_panel/types/ControlProps";

const controlProps: ControlProps = {
  isEditMode: true,
  theme: "dark",
  results: {},
  updateCtrlValue: jest.fn(),
  attachParamsToCtrl: jest.fn(),
  removeCtrl: jest.fn(),
  setCurrentInput: jest.fn(),
  setOpenEditModal: jest.fn(),
};
// mock `useFlowChartState` hook
jest.mock("@src/hooks/useFlowChartState");
// mock `Control` component
jest.mock("@src/feature/controls_panel/views/Control", () => ({
  __esModule: true,
  default: jest.fn(() => <div data-testid="control"></div>),
}));

describe("ControlGrid.tsx", () => {
  it("renders the ControlGrid component.", () => {
    const { container } = render(<ControlGrid controlProps={controlProps} />);
    expect(container).toMatchSnapshot();
  });
});
