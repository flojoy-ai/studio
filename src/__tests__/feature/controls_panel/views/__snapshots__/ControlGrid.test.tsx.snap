// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlGrid.tsx renders the ControlGrid component. 1`] = `
<div>
  <div
    data-testid="react-grid"
  >
    <div
      data-grid="[object Object]"
      style="background-color: rgb(25, 25, 25); border-radius: 16px;"
    >
      <div
        data-testid="control"
      />
    </div>
  </div>
</div>
`;
